services/venice-server/src/main/java/com/linkedin/venice/stats/ServerQuotaUsageStats.java

```diff
@@ -6,38 +6,44 @@
 import io.tehuti.metrics.stats.AsyncGauge;
 import io.tehuti.metrics.stats.Count;
 import io.tehuti.metrics.stats.Rate;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 
 
 /**
  * for measuring requests and quota rejections for each store
  */
 public class ServerQuotaUsageStats extends AbstractVeniceStats {
   private final Sensor requestedQPS; // requested query per second
-  private final Sensor requestedKPS; // requested key per second
+  private final AtomicReference<Sensor> requestedKPS = new AtomicReference<>(); // requested key per second
   private final Sensor rejectedQPS; // rejected query per second
   private final Sensor rejectedKPS; // rejected key per second
   private final Sensor allowedUnintentionallyKPS; // allowed KPS unintentionally due to error or insufficient info
-  private final Sensor usageRatioSensor; // requested qps divided by nodes quota responsibility
+  private final Sensor usageRatioSensor; // requested kps divided by nodes quota responsibility
+  private final Sensor nodeResponsibilitySensor; // kps assigned to the node based on store quota and replica assignment
+  private static final String REQUESTED_KPS_SENSOR_NAME = "quota_request_key_count";
 
-  private long nodeQpsResponsibility = 0;
+  private final AtomicLong nodeKpsResponsibility = new AtomicLong(0);
 
```

Comment:
I have some question about the calculation of request KPS/quota and usage.
Here is my assumption:
1. Request KPS will cover all the KPS for the given store, which can from the current version (mostly) and backup version(during swap time as fast-client version swap is not atomic and has delays and the switch can happen at different times in different nodes).
2. KPS quota will be bounded to a specific version.
3. Quota is the total kps (can from mixed versions) / total quota for a specific version.

If the above assumption is correct, it is possible that the quota usage during the version swap can still be weird and if some fast client nodes stick to the backup version for longer period unexpectedly, the ratio can be hard to reason.
Ideally, we should align the version between these two metrics, so that the quota usage ratio will be more accurate.
Here is an idea:
1. Maintain a separate KPS metric for backup and current version.
2. Quota will be calculated based on the current version.
3. Calculate the quota usage ratio by the current version KPS / current version quota.
4. We can emit the KPS metric for backup version, and we can setup alert against this metric if this value doesn't go down quickly (which indicates some fast-client nodes are not flipping and it will be great if we can log the problematic nodes as well, for example, if the backup version KPS doesn't go down to 0 within 3 mins, we will log some sample request node hostnames periodically for debugging purpose).
WDYT?

Reply:
That's a good observation, I do think the recommendation is worth pursing since it will also give us some additional visibility on top of a more accurate usage ratio. Let me make the changfe.